aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1315,81 +1315,255 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+        }
+        *ptr_nneg_index++ = idx;
+      }
+    }
 
-    auto cpu_dim_indices = indices[dim].to(c10::kCPU).contiguous();
-    int64_t* cpu_dim_indices_ptr = cpu_dim_indices.data_ptr<int64_t>();
-    auto cpu_index = index.to(c10::kCPU).contiguous();
-    int64_t* cpu_index_ptr = cpu_index.data_ptr<int64_t>();
-    std::vector<int64_t> zindices;
-    std::vector<int64_t> iindices;
-    int64_t new_nnz = 0;
-    for (const auto i : c10::irange(new_sizes[dim])) {
-      int64_t idx = cpu_index_ptr[i];
-      if (idx < -size || idx >= size) {
-        TORCH_CHECK_INDEX(false, "index_select(): index contains ", idx, " that is out of range for tensor of size ",
-                   self.sizes(), " at dimension ", dim);
+    // Much faster than at::unique_dim
+    const auto unique_with_counts = [](
+        const Tensor& t, int64_t len, bool is_sorted = false
+    ) -> std::tuple<Tensor, Tensor, Tensor, int64_t> {
+      // t_unique will be modified to hold unique values in-place
+      Tensor t_unique, t_idx;
+      if (!is_sorted) {
+        std::tie(t_unique, t_idx) = at::sort(t);
+      }
+      else {
+        // faster than t_unique = t.clone()
+        t_unique = at::empty_like(t);
+        t_unique.copy_(t);
+        t_idx = at::arange(0, t.numel(), t.options());
       }
-      if (idx < 0) {
-        idx += size;
+
+      auto t_counts = at::ones_like(t_unique);
+      int64_t len_unique;
+      {
+        auto* ptr_counts = t_counts.data_ptr<int64_t>();
+        auto* first = t_unique.data_ptr<int64_t>();
+        auto* last = first + len;
+        auto* curr = first;
+        auto* counts = ptr_counts;
+        while (++first != last) {
+          if (*curr != *first) {
+            ++counts;
+            if (++curr != first) {
+              *curr = *first;
+            }
+          }
+          else {
+            ++(*counts);
+          }
+        }
+        len_unique = counts - ptr_counts + 1;
       }
-      for (const auto j : c10::irange(nnz)) {
-        int64_t jdx = cpu_dim_indices_ptr[j];
-        if (idx == jdx) {
-          new_nnz++;
-          iindices.push_back(i);
-          zindices.push_back(j);
+
+      return std::make_tuple(t_idx, t_unique, t_counts, len_unique);
+    };
+
+    Tensor dim_sort_indices, dim_indices_unique, dim_indices_counts;
+    int64_t n_unique_dim_indices;
+    std::tie(dim_sort_indices, dim_indices_unique, dim_indices_counts, n_unique_dim_indices)
+      = unique_with_counts(
+          indices[dim], nnz,
+          // if self is coalesced and dim == 0,
+          // indices[dim] is already sorted
+          /*is_sorted=*/self.is_coalesced() && dim == 0
+        );
+
+    Tensor sel_sort_indices, sel_indices_unique, sel_indices_counts;
+    int64_t n_unique_sel_indices;
+    std::tie(sel_sort_indices, sel_indices_unique, sel_indices_counts, n_unique_sel_indices)
+      = unique_with_counts(nneg_index, index_len);
+
+    // index intersection is aka `set(indices[dim]) - set(index)`.
```

Comment:
index intersection of two sparse Tensors would be a generally useful algorithm for binary operations etc.